### ISSUE TITLE
System.IO.EndOfStreamException: SharedQueue closed

### ISSUE THREAD

**Issue by mattdone01** [Particular/NServiceBus.RabbitMQ#84]
I keep getting the exception System.IO.EndOfStreamException: SharedQueue closed. This happens in the RabbitMQ.Client when it tries to dequeue a closed queue. I would prefer to have no dequeueTimeout and set it to -1 in the connection string. This should fix my issue. Relevant detail is below. I get hundreds of these every day.

The stack trace
```
NServiceBus.Transports.RabbitMQ.RabbitMqDequeueStrategy   : Failed to receive messages from saturn.customers.input.TimeoutsDispatcher | System.AggregateException System.AggregateException: One or more errors occurred. ---> System.IO.EndOfStreamException: SharedQueue closed
   at RabbitMQ.Util.SharedQueue`1.Dequeue(Int32 millisecondsTimeout, T& result)
   at NServiceBus.Transports.RabbitMQ.RabbitMqDequeueStrategy.DequeueMessage(QueueingBasicConsumer consumer, Int32 dequeueTimeout) in c:\BuildAgent\work\ef98ad7376e3379a\src\NServiceBus.RabbitMQ\RabbitMqDequeueStrategy.cs:line 257
   at NServiceBus.Transports.RabbitMQ.RabbitMqDequeueStrategy.ConsumeMessages(Object state) in c:\BuildAgent\work\ef98ad7376e3379a\src\NServiceBus.RabbitMQ\RabbitMqDequeueStrategy.cs:line 247
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.IO.EndOfStreamException: SharedQueue closed
   at RabbitMQ.Util.SharedQueue`1.Dequeue(Int32 millisecondsTimeout, T& result)
   at NServiceBus.Transports.RabbitMQ.RabbitMqDequeueStrategy.DequeueMessage(QueueingBasicConsumer consumer, Int32 dequeueTimeout) in c:\BuildAgent\work\ef98ad7376e3379a\src\NServiceBus.RabbitMQ\RabbitMqDequeueStrategy.cs:line 257
   at NServiceBus.Transports.RabbitMQ.RabbitMqDequeueStrategy.ConsumeMessages(Object state) in c:\BuildAgent\work\ef98ad7376e3379a\src\NServiceBus.RabbitMQ\RabbitMqDequeueStrategy.cs:line 247
   at System.Threading.Tasks.Task.Execute()<---
```

On line 198 of RabbitMQ.Util.SharedQueue  there is some code that says
if (millisecondsTimeout == -1)

This is the dequeue timeout. If the dequeueTiemout is set to -1 RabbitMQClient bypasses any timeout code when dequeuing.

**Comment by andreasohlund** [comment]
What's the behavior when Timeout is -1 ? (return right away if no message or wait indefinitely?)

**Comment by mattdone01** [comment]
Hi,
  I think we figured it out. We were load balancing three RabbitMQ clustered servers via a round robin loadbalancer (netscaler). We had a DNS of rabbitmqcluster.internal.net pointing to the virtual IP. We think what was happening was that the load balancer was moving us onto another server during the dequeue and this caused all the messages we were seeing. Instead of the DNS point to the VIP we are now using the actual DNS hostnames in the connection string. Although we don't get the load balanced across server which wasn't really that important we still get the redundancy.

I also have another issue I raised on Stack overflow in regards to load balancing. Would love to chat to you about it. Thanks Andreas.

**Comment by andreasohlund** [comment]
Link to the SO issue?

**Comment by andreasohlund** [comment]
Do you still believe this issue is still a valid improvement or should we close it?

**Comment by mattdone01** [comment]
Well there is provision for it in the RabbitMQ client. if
dequeueTimeout==-1 Probably not worth the hassle though.

The SO issue I was talking about is here
http://stackoverflow.com/questions/29642509/competing-consumers-in-rabbitmq-via-nservicebus
Thanks alot.

**Comment by andreasohlund** [comment]
Side note: Did you figure out the solution from this doco? 

http://docs.particular.net/nservicebus/rabbitmq/configuration-api

Does that doco make sense? anything we could improve?

**Comment by andreasohlund** [comment]
>Well there is provision for it in the RabbitMQ client. if
dequeueTimeout==-1 Probably not worth the hassle though.

I'll need to research this a bit more since we're in a tight loop so this could case CPU churn. I'll keep this one open until I've tested it

**Comment by mattdone01** [comment]
Ok,
  I do think you need to mention something about load balancing and
clustering and how it may effect the queue connection. It caused me a lot
of head ache trying to figure it out. I am still not 100% sure if this is
the issue but I will let you know if it was. I may also investigate
enabling sticky sessions on the load balancer, that may help.

**Comment by andreasohlund** [comment]
You mean the fact that we connect to the first one in the list always?

**Comment by mattdone01** [comment]
No I mean lad balancing in general. Typically when you setup a messaging
cluster you can load balance it behind a F5 or netscaller hardware load
balancer. Most cloud providers do this. I know AWS is doing this. So you
will see errors like this from cloud deployments or large enterprise
deployments like mine.

Chaining 3 host names in a connection string like NSB does isn't balancing
the load it is just managing redundancy. It might be a future enhancement
for you guys.

MD

**Comment by andreasohlund** [comment]
We're deprecating the multi connection support in 3.0.0 so this will be covered. Ie we'll add basic recommendations to use a load balancer in front of the cluster

**Comment by SzymonPobiega** [comment]
Prerequisite for #91 because the multi-host feature was a workaround for the problem described here

**Comment by andreasohlund** [comment]
I think it would still work, we're not removing the re-connections but instead it would reconnect via the load balancer which hopefully would forward the request to another instance should one of them have keeled over?

**Comment by bording** [comment]
The 4.0 message pump has been designed in a way that should completely remove the possibility of this error from occurring.

I'm going to consider this one fixed in 4.0.
